x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -853,75 +840,76 @@ private void putTransientIfNonExisting(String key, Object value) {
         }
     }
 
-    private ElasticsearchSecurityException denialException(
+    private ElasticsearchSecurityException runAsDenied(Authentication authentication, String action) {
+        return denialException(authentication, action, () -> AuthorizationDenialMessages.runAsDenied(authentication, action), null);
+    }
+
+    private ElasticsearchSecurityException bulkItemActionDenied(
         Authentication authentication,
         String action,
         TransportRequest request,
-        Exception cause
+        String resolvedIndex
     ) {
-        return denialException(authentication, action, request, null, cause);
+        return actionDenied(
+            authentication,
+            action,
+            request,
+            IndexAuthorizationResult.getFailureDescription(List.of(resolvedIndex), restrictedIndices),
+            null
+        );
     }
```

Comment:
This method does not look necessary to me. It's a very thin wrapper for `actionDeined` and the only thing it really does is to calcualte the `context` String. But to do that it has to introduce a `String resovledIndex` argument which is an anomaly in the series of `xxxDenied` methods. I'd rather stick to the existing five arguments (which is already plenty) and directly invoke `actionDenied` at the caller and remove this indrection.

Reply:
Makes sense! I was thinking about moving closer to hiding the details of how messages are created but the implementation falls flat of that so I'm for keeping it simple and less cluttered.